wcfsetup/install/files/lib/data/reaction/ReactionAction.class.php

```diff
@@ -0,0 +1,343 @@
+		}
+		if ($this->parameters['targetType'] == 'received') {
+			$likeList->getConditionBuilder()->add("like_table.objectUserID = ?", [$this->parameters['userID']]);
+		}
+		else {
+			$likeList->getConditionBuilder()->add("like_table.userID = ?", [$this->parameters['userID']]);
+		}
+		$likeList->getConditionBuilder()->add("like_table.reactionTypeID = ?", [$this->parameters['reactionTypeID']]);
+		$likeList->readObjects();
+		
+		if (empty($likeList)) {
+			return [];
+		}
+		
+		// parse template
+		WCF::getTPL()->assign([
+			'likeList' => $likeList
+		]);
+		
+		return [
+			'lastLikeTime' => $likeList->getLastLikeTime(),
+			'template' => WCF::getTPL()->fetch('userProfileLikeItem')
+		];
+	}
+	
+	/**
+	 * Copies likes from one object id to another.
+	 */
+	public function copy() {
+		$sourceObjectType = ObjectTypeCache::getInstance()->getObjectTypeByName('com.woltlab.wcf.like.likeableObject', $this->parameters['sourceObjectType']);
+		$targetObjectType = ObjectTypeCache::getInstance()->getObjectTypeByName('com.woltlab.wcf.like.likeableObject', $this->parameters['targetObjectType']);
+		
+		//
+		// step 1) get data
+		//
+		
+		// get like object
+		$sql = "SELECT	*
+			FROM	wcf".WCF_N."_like_object
+			WHERE	objectTypeID = ?
+				AND objectID = ?";
+		$statement = WCF::getDB()->prepareStatement($sql);
+		$statement->execute([
+			$sourceObjectType->objectTypeID,
+			$this->parameters['sourceObjectID']
+		]);
+		$row = $statement->fetchArray();
+		
+		// no (dis-)likes at all
+		if ($row === false) {
+			return;
+		}
+		
+		unset($row['likeObjectID']);
+		$row['objectTypeID'] = $targetObjectType->objectTypeID;
+		$row['objectID'] = $this->parameters['targetObjectID'];
+		$newLikeObject = LikeObjectEditor::create($row);
+		
+		//
+		// step 2) copy
+		//
+		
+		$sql = "INSERT INTO	wcf".WCF_N."_like
+					(objectID, objectTypeID, objectUserID, userID, time, likeValue, reactionTypeID)
+			SELECT		".$this->parameters['targetObjectID'].", ".$targetObjectType->objectTypeID.", objectUserID, userID, time, likeValue, reactionTypeID
+			FROM		wcf".WCF_N."_like
+			WHERE		objectTypeID = ?
+					AND objectID = ?";
+		$statement = WCF::getDB()->prepareStatement($sql);
+		$statement->execute([
+			$sourceObjectType->objectTypeID,
+			$this->parameters['sourceObjectID']
+		]);
+		
+		//
+		// step 3) update owner
+		//
+		
+		if ($newLikeObject->objectUserID) {
+			$sql = "SELECT	COUNT(*) as count, like_table.reactionTypeID
+				FROM	wcf".WCF_N."_like like_table
+				WHERE	like_table.objectTypeID = ?
+					AND like_table.objectID = ?
+				GROUP BY like_table.reactionTypeID";
+			$statement = WCF::getDB()->prepareStatement($sql);
+			$statement->execute([
+				$targetObjectType->objectTypeID,
+				$this->parameters['targetObjectID']
+			]);
+			
+			$updateValues = [
+				'positive' => 0,
+				'neutral' => 0, 
+				'negative' => 0
+			];
+			while ($row = $statement->fetchArray()) {
```

Comment:
I'm not sure whether this should be counted in PHP. It could possibly be replaced by three queries, each counting the number of non-disabled reactions for a given sentiment.